### PR TITLE
Expose customizable URLRequest

### PR DESCRIPTION
**Pull request by BorysKasianenko-okta**
### Problem Analysis (Technical)

This is a requirement from our customer, they want to have an easy way to add properties to the URLRequest.

This PR based on #236 . We want to leave `LL-OKTA-314365-URLRequest` branch for testing purposes and finish the work here.  
Changes:

- Addressed previous PR comments;
- Added unit tests;
- Updated docs and readme.

### Solution (Technical)

Introduce a new protocol
```
@protocol OktaOidcHTTPProtocol
(void)willSendRequest: (nullable NSURLRequest *)request;
(void)didReceiveResponse: (nullable NSURLResponse *)response;
@end 
```

### Affected Components

-

### Steps to reproduce:

-

### Tests

 - `OktaOidcRestApiTests`
 - `OIDAuthorizationServiceRequestDelegateTests`

@IldarAbdullin-okta @lihaoli-okta @umangshah-okta @stevelind-okta @okta/ios

### CODE REVIEW

File: Okta/OktaOidc/OktaOidc.swift [hunk 1]
```diff
@@ -25,16 +25,21 @@ public class OktaOidc: NSObject {
 
     @objc public func authenticate(withSessionToken sessionToken: String,
                                    callback: @escaping ((OktaOidcStateManager?, Error?) -> Void)) {
-        OktaOidcAuthenticateTask(config: configuration, oktaAPI: OktaOidcRestApi())
-        .authenticateWithSessionToken(sessionToken: sessionToken) { authState, error in
+        OktaOidcAuthenticateTask(config: configuration,
```
Comment (IldarAbdullin-okta) [inline Okta/OktaOidc/OktaOidc.swift:28]:
```suggestion
        let task = OktaOidcAuthenticateTask(...)
        task. authenticateWithSessionToken(...)
```

File: Okta/OktaOidc/OktaOidc.swift [hunk 2]
```diff
@@ -45,14 +50,17 @@ public class OktaOidc: NSObject {
...
+        task.signIn(delegate: configuration.requestCustomizationDelegate) { [weak self] authState, error in
             defer { self?.currentUserSessionTask = nil }
             guard let authState = authState else {
                 callback(nil, error)
                 return
             }
             
             let authStateManager = OktaOidcStateManager(authState: authState)
+            if let delegate = self?.configuration.requestCustomizationDelegate {
+                authStateManager.restAPI = OktaOidcRestApi(delegate: delegate)
```
Comment (IldarAbdullin-okta) [inline Okta/OktaOidc/OktaOidc.swift:62]: Why just not:
```
authStateManager.restAPI. requestCustomizationDelegate = delegate
```
I would just avoid recreating `OktaOidcRestApi` outside of `OktaOidcStateManager` as you don't know all specifics how `OktaOidcRestApi` is created and configured inside `OktaOidcStateManager`

File: Okta/OktaOidc/OktaOidc.swift [hunk 3]
```diff
@@ -25,16 +25,21 @@ public class OktaOidc: NSObject {
...
+                                                                                                                          delegate: configuration.requestCustomizationDelegate,
+                                                                                                            callback: { (authState, error) in
             guard let authState = authState else {
                 callback(nil, error)
                 return
             }
             
             let authStateManager = OktaOidcStateManager(authState: authState)
+            if let delegate = self.configuration.requestCustomizationDelegate {
+                authStateManager.restAPI = OktaOidcRestApi(delegate: delegate)
```
Comment (IldarAbdullin-okta) [inline Okta/OktaOidc/OktaOidc.swift:39]: Please remove `delegate` from `OktaOidcRestApi` constructor and use property injection

File: README.md [hunk 4]
```diff
@@ -167,6 +167,16 @@ if #available(iOS 13.0, *) {
 ```
 ***Note*** Flag is available on iOS 13 and above versions
 
+### Modify network requests
```
Comment (IldarAbdullin-okta) [inline README.md:170]: Please add this to table of contents

Comment (IldarAbdullin-okta) [reply README.md:170]: I would move this to the end of the readme, just before Known-Issues

File: README.md [hunk 5]
```diff
@@ -167,6 +167,16 @@ if #available(iOS 13.0, *) {
...
+
+You can track and modify network requests throughout OktaOidc by passing object that conforms to `OktaNetworkRequestCustomizationDelegate` protocol to the `requestCustomizationDelegate` property of `OktaOidcConfig` instance.
+
+```swift
+let configuration = OktaOidcConfig(with: {YourOidcConfiguration})
+configuration.requestCustomizationDelegate = {YourDelegateInstance}
```
Comment (IldarAbdullin-okta) [inline README.md:176]: Please add example how to use delegate methods

**Review by IldarAbdullin-okta — CHANGES_REQUESTED** [review]
I didn't find changes in `OIDAuthorizationService+Okta.swift` category and unit tests for the category

File: Okta/OktaOidc/OktaOidc.swift (hunk 1 above)
Comment (BorysKasianenko-okta) [reply Okta/OktaOidc/OktaOidc.swift]: Fixed, thanks.

File: Okta/OktaOidc/OktaOidc.swift (hunk 2 above)
Comment (BorysKasianenko-okta) [reply Okta/OktaOidc/OktaOidc.swift]: I've removed recreating `OktaOidcRestApi` here, but there are few other places in this class where we creating `OktaOidcRestApi` object. Should we consider to change this as well?

File: Okta/OktaOidc/OktaOidc.swift (hunk 3 above)
Comment (BorysKasianenko-okta) [reply Okta/OktaOidc/OktaOidc.swift]: I've removed delegate from init, thanks.

File: README.md (hunk 4 above)
Comment (BorysKasianenko-okta) [reply README.md]: Done

File: README.md (hunk 5 above)
Comment (BorysKasianenko-okta) [reply README.md]: Done

File: README.md [hunk 6]
```diff
@@ -452,6 +453,40 @@ bash ./scripts/build-and-test.sh
 
 **Note:** *You may need to update the emulator device to match your Xcode version*
 
+## Modify network requests
+
+You can track and modify network requests throughout OktaOidc by passing object that conforms to `OktaNetworkRequestCustomizationDelegate` protocol to the `requestCustomizationDelegate` property of `OktaOidcConfig` instance.
```
Comment (IldarAbdullin-okta) [inline README.md:458]: @stevelind-okta , please help to validate spelling. Thank you! ;)

**Review by IldarAbdullin-okta — APPROVED** [review]
LGTM!

File: README.md (hunk 6 above)
Comment (stevelind-okta) [reply README.md:458]: Suggested change:
 
You can track and modify network requests made by `OktaOidc`. In order to do this, create an object conforming to the `OktaNetworkRequestCustomizationDelegate` protocol and set it to the `requestCustomizationDelegate` property on an `OktaOidcConfig` instance.

Comment (BorysKasianenko-okta) [reply README.md]: Thank you, Steve! I'll update the readme.

**Pull request merged**